pygfx/renderers/wgpu/shader/base.py

```diff
@@ -0,0 +1,285 @@
+"""
+Implements the base shader class. The shader is responsible for providing the
+WGSL code, as well as providing the information to connect it to the resources
+(buffers and textures) and some details on the pipeline and rendering.
+"""
+
+from ....resources import Buffer, Texture
+from ..engine.utils import (
+    GfxSampler,
+    GfxTextureView,
+    to_vertex_format,
+    to_texture_format,
+    hash_from_value,
+)
+from ..engine.binding import Binding
+from .resolve import resolve_varyings, resolve_depth_output
+from .definitions import BindingDefinitions
+from .templating import apply_templating
+
+
+class ShaderInterface:
+    """Define what a shader object must look like from the pov from the pipeline."""
+
+    def __init__(self):
+        self._hash = None
+
+    def lock_hash(self):
+        self._hash = self._get_hash()
+
+    def unlock_hash(self):
+        self._hash = None
+
+    @property
+    def hash(self):
+        """A hash of the current state of the shader. If the hash changed, it's likely that the shader changed."""
+        if self._hash is None:
+            return self._get_hash()
+        else:
+            return self._hash
+
+    def _get_hash(self):
+        raise NotImplementedError()
+
+    def generate_wgsl(self, **kwargs):
+        raise NotImplementedError()
+
+    def get_bindings(self, wobject, shared):
+        """Subclasses must return a dict describing the buffers and
+        textures used by this shader.
+
+        The result must be a dict of dicts with binding objects
+        (group_slot -> binding_slot -> binding)
+        """
+        return {
+            0: {},
+        }
+
+    def get_pipeline_info(self, wobject, shared):
+        """Subclasses must return a dict describing pipeline details.
+
+        Fields for a compute shader: empty
+
+        Fields for a render shader:
+          * "cull_mode"
+          * "primitive_topology"
+        """
+        return {
+            "primitive_topology": 0,
+            "cull_mode": 0,
+        }
+
+    def get_render_info(self, wobject, shared):
+        """Subclasses must return a dict describing render details.
+
+        Fields for a compute shader:
+          * "indices" (3 ints)
+
+        Fields for a render shader:
+          * "render_mask"
+          * "indices" (list of 2 or 4 ints).
+        """
+        return {
+            "indices": (1, 1),
+            "render_mask": 0,
+        }
+
+
+class BaseShader(ShaderInterface):
```

Comment:
This class is the combination of what was previously `BaseShader` and `WorldObjectShader`. Other changes to the code are mostly confined to the `generate_wgsl()` method.